File: rubygems/pray-cli/lib/pray/apply_report.rb
# frozen_string_literal: true

module Pray
  module ApplyReport
    module_function

    def local_summary_lines(previous, project)
      previous_checksums = previous_local_checksums(previous)
      project.local_files.filter_map do |local|
        next if local.content.empty? && local.optional

        checksum = local.source_checksum
        previous_checksum = previous_checksums[local.manifest_path]
        if previous_checksum.nil?
          "Installing #{local.manifest_path} (#{checksum})"
        elsif previous_checksum == checksum
          "Using #{local.manifest_path} (#{checksum} checked)"
        else
          "Updating #{local.manifest_path} (#{checksum} was #{previous_checksum})"
        end
      end
    end

    def outdated_local_lines(previous, project)
      previous_checksums = previous_local_checksums(previous)
      project.local_files.filter_map do |local|
        next if local.content.empty? && local.optional

        checksum = local.source_checksum
        previous_checksum = previous_checksums[local.manifest_path]
        if previous_checksum && previous_checksum != checksum
          "#{local.manifest_path} #{previous_checksum} -> #{checksum}"
        elsif previous && previous_checksum.nil?
          "#{local.manifest_path} (new) -> #{checksum}"
        end
      end
    end

    def previous_local_checksums(previous)
      return {} unless previous

      previous.managed_span.each_with_object({}) do |span, checksums|
        next unless span.package == LOCAL_EMBED_PACKAGE

        checksums[span.export] = span.source_checksum
      end
    end
  end
end

File: npmjs/pray-cli/src/cli/apply-report.ts
import type { Lockfile } from "../lockfile/types.js";
import { LOCAL_EMBED_PACKAGE } from "../lockfile/types.js";
import type { ResolvedProject } from "../resolve/types.js";

export function localSummaryLines(
  previous: Lockfile | undefined,
  project: ResolvedProject,
): string[] {
  const previousChecksums = previousLocalChecksums(previous);
  const lines: string[] = [];
  for (const local of project.localFiles) {
    if (local.content === "" && local.optional) {
      continue;
    }
    const checksum = local.sourceChecksum;
    const previousChecksum = previousChecksums.get(local.manifestPath);
    if (previousChecksum === undefined) {
      lines.push(`Installing ${local.manifestPath} (${checksum})`);
    } else if (previousChecksum === checksum) {
      lines.push(`Using ${local.manifestPath} (${checksum} checked)`);
    } else {
      lines.push(
        `Updating ${local.manifestPath} (${checksum} was ${previousChecksum})`,
      );
    }
  }
  return lines;
}

export function outdatedLocalLines(
  previous: Lockfile | undefined,
  project: ResolvedProject,
): string[] {
  const previousChecksums = previousLocalChecksums(previous);
  const lines: string[] = [];
  for (const local of project.localFiles) {
    if (local.content === "" && local.optional) {
      continue;
    }
    const checksum = local.sourceChecksum;
    const previousChecksum = previousChecksums.get(local.manifestPath);
    if (previousChecksum !== undefined && previousChecksum !== checksum) {
      lines.push(`${local.manifestPath} ${previousChecksum} -> ${checksum}`);
    } else if (previous !== undefined && previousChecksum === undefined) {
      lines.push(`${local.manifestPath} (new) -> ${checksum}`);
    }
  }
  return lines;
}

function previousLocalChecksums(
  previous: Lockfile | undefined,
): Map<string, string> {
  const checksums = new Map<string, string>();
  if (!previous) {
    return checksums;
  }
  for (const span of previous.managed_span) {
    if (span.package === LOCAL_EMBED_PACKAGE) {
      checksums.set(span.export, span.source_checksum);
    }
  }
  return checksums;
}
